### ISSUE TITLE
Can not Understand CMC Code 

### ISSUE THREAD

**Issue by yxxxxxxxx** [huanghoujing/AlignedReID-Re-Production-Pytorch#17]
Hello! I am a student who is using your code calculating the CMC. But recently I'm stuck with your code when calculating the CMC in aligned_reid/utils/metrics.py  :
matches = (gallery_ids[indices] == query_ids[:, np.newaxis]) 
I wonder what this line returns and what the gallert_ids[indices] and the query_ids[:,np.newaxis] respectively
looks like. I would appreciate it if you could give me an simple example to explain this line! thank you very much!

**Comment by huanghoujing** [comment]
The core part of the metric code is from open-reid. 

Here `gallery_ids` has shape `[num_gallery]`, `indices` has shape `[num_query, num_gallery]`, thus `gallery_ids[indices]` has shape `[num_query, num_gallery]`. Each row of `gallery_ids[indices]` is for a query, containing the sorted gallery ids. 

`query_ids` has shape `[num_query]`, thus `query_ids[:, np.newaxis]` has shape `[num_query, 1]`. 

`matches = (gallery_ids[indices] == query_ids[:, np.newaxis])` has shape `[num_query, num_gallery]`. Each row of `matches` contains `True` and `False` values, where `True` means same id as query.

The part confusing you may be this `gallery_ids[indices]`. May the following example helps:
```python
from __future__ import print_function
import numpy as np

gallery_ids = np.array([1, 2, 4, 3, 1])
indices = np.array([[0, 1, 3, 4, 2],
                    [1, 3, 0, 2, 4],
                    [1, 0, 3, 4, 2]])
print(gallery_ids[indices])
```

The result is 
```
[[ 5  8  7  3 12]
 [ 8  7  5 12  3]
 [ 8  5  7  3 12]]
```

**Comment by yxxxxxxxx** [comment]
Thank you for taking time to solve my problem, now I understand your meaning but when I run the following code, the matches always return false.
`
from __future__ import print_function
import numpy as np

gallery_ids = np.array([1, 2, 4, 3, 1])
query_ids = np.array([1, 2, 4, 5, 1])
indices = np.array([[0, 1, 3, 4, 2],
                    [1, 3, 0, 2, 4],
                    [1, 0, 3, 4, 2]])
matches = (gallery_ids[indices] == query_ids[:, np.newaxis])
print(matches)
`
the result is only a False. Is this problem caused by my Numpy Version or somgthing else is wrong.thank you!

**Comment by huanghoujing** [comment]
I made a mistake in the previous answer. I have updated it now.

**Comment by yxxxxxxxx** [comment]
Ok I have noticed your error and that's all right. But I found another problem as commented just now ,could you explian this problem for me? Thank you very much.

**Comment by huanghoujing** [comment]
```python
from __future__ import print_function
import numpy as np

gallery_ids = np.array([1, 2, 4, 3, 1])
query_ids = np.array([1, 2, 4, 5, 1])
indices = np.array([[0, 1, 3, 4, 2],
                    [1, 3, 0, 2, 4],
                    [1, 0, 3, 4, 2]])

print(gallery_ids[indices])

print(gallery_ids[indices] == query_ids)
```

It outputs:

```
[[1 2 3 1 4]
 [2 3 1 4 1]
 [2 1 3 1 4]]
[[ True  True False False False]
 [False False False False  True]
 [False False False False False]]
```

**Comment by yxxxxxxxx** [comment]
Ok. so I should remove the [:, np.newaxis] behind the query_ids?but the code doesn't remove it.

**Comment by huanghoujing** [comment]
I gave a wrong example in which the dimension of `indices` is not consistent with `[num_query, num_gallery]`.

The following one should be practical:

```python
from __future__ import print_function
import numpy as np

gallery_ids = np.array([1, 2, 4, 3, 1])
query_ids = np.array([1, 2, 4])
indices = np.array([[0, 1, 3, 4, 2],
                    [1, 3, 0, 2, 4],
                    [1, 0, 3, 4, 2]])

print(gallery_ids[indices])

print(gallery_ids[indices] == query_ids[:, np.newaxis])
```

It outputs:

```
[[1 2 3 1 4]
 [2 3 1 4 1]
 [2 1 3 1 4]]
[[ True False False  True False]
 [ True False False False False]
 [False False False False  True]]
```

**Comment by yxxxxxxxx** [comment]
Thank you very much! I understand this problem!

**Comment by huanghoujing** [comment]
You're welcome. The high dimension data, `argsort`, indexing, etc also confuse me at times.

**Comment by yxxxxxxxx** [comment]
Sorry to disturb you again but I am not confused with the code 

`delta = 1. / (len(index) * repeat)`

and what the delta stands for?for the Market1501 dataset the value of delta always equals to 1,so why add delta in the code below
`ret[i, k - j] += delta` 
thank you!

**Comment by huanghoujing** [comment]
`delta` is for the option `first_match_break`. When analyzing line `delta = 1. / (len(index) * repeat)`, you can just consider `repeat == 1`, and it becomes `delta = 1. / len(index)`. The `* repeat` here is just a way of implementing *averaging over repeats*.
- If `first_match_break` is `True`, the resulting CMC curve for a query is a step function turning from `0` to `1` at the point where the first valid match in gallery list appears.
- If `first_match_break` is `False`, the resulting CMC curve for a query is a staircase function going up by `delta` wherever a valid match in gallery list appears. This function finally goes up to `1`. The `delta` for a query is `1 / num-of-valid-match`.

In the case of Market1501, `first_match_break == True`, thus `delta` is not used. The following three lines instead of `ret[k - j] += delta` are executed.

```python
if first_match_break:
    ret[k - j] += 1
    break
```

**Comment by yxxxxxxxx** [comment]
Hello houjing, now I am reproducing your code on the cuhk03 dataset, now I understand the new protocol which split the training set and testing set  which consist of 767 identities and 700 identities respectively. But I want to use the papers protocol which split the dataset once for training and testing, and the gallery includes 200 identities. I have noticed that you use the re_ranking_train_test_split.pkl to produce the partition file. So I wonder if I want to produce the paper's protocol,how or where you get this .pkl file  so that I can imitate to get a new .pkl file, for example if there is a script to produce it. thank you!

**Comment by huanghoujing** [comment]
Sorry, I do not have this 1267/200 train/test split file. 

To generate a `train_test_partition_file` for [this function](https://github.com/huanghoujing/AlignedReID-Re-Production-Pytorch/blob/bc9e1bc51a0112ecb813cf81af151577e2e3c58a/script/dataset/transform_cuhk03.py#L68), you have to know the structure of this `.pkl` file. It is a dictionary (denoted by `train_test_partition` in the following) having two keys `detected` and `labeled`. The following takes the `detected` subset as an example. `train_test_partition['detected']` is a dictionary with keys `train_im_names`, `query_im_names` and `gallery_im_names`.

If you want to generate a 1267/200 train/test split file, `train_test_partition['detected']['train_im_names']` should contain image names of the 1267 ids and `train_test_partition['detected']['query_im_names']` + `train_test_partition['detected']['gallery_im_names']` should contain image names of the rest 200 ids. Which 1267 ids to select as training ids, and how the test set is split into query and gallery sets are up to you.
